### PR TITLE
Potion NamespacedKey. again...

### PR DESCRIPTION
**Pull request by DaRacci**
Continued from https://github.com/pl3xgaming/Purpur/pull/710
Sorry for closing and reopening a new pr again, i was having issue with my previous fork and I'm not too experienced with pr and fork stuff so i deleted it and forked again to make the changes needed.

I've had these changes in use on a production server for about a month or so with no more issues so it should be all good with plugins and what not.

- I've fixed an issue that caused  serialized potion effects to cause null issues etc
- Now custom potion effects on CraftMetaPotion have the namespacedKeys too
- No longer serialize null namespaces for potions
- Added some annotations
- Probably other misc stuff too

### CODE REVIEW

File: patches/api/0045-Potion-NamespacedKey.patch [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+@@ -242,9 +300,9 @@ public class PotionEffect implements ConfigurationSerializable {
+      * @return color of this potion's particles. May be null if the potion has no particles or defined color.
+      * @deprecated color is not part of potion effects
+      */
+-     @Deprecated
+-     @Nullable
+-     @Contract("-> null")
++    @Deprecated
++    @Nullable
++    @Contract("-> null")
```
Comment (Euphillya) [inline patches/api/0045-Potion-NamespacedKey.patch]: Restore indentation

File: patches/api/0045-Potion-NamespacedKey.patch [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+-        this(getEffectType(map), getInt(map, DURATION), getInt(map, AMPLIFIER), getBool(map, AMBIENT, false), getBool(map, PARTICLES, true), getBool(map, ICON, getBool(map, PARTICLES, true)));
++        this(getEffectType(map), getInt(map, DURATION), getInt(map, AMPLIFIER), getBool(map, AMBIENT, false), getBool(map, PARTICLES, true), getBool(map, ICON, getBool(map, PARTICLES, true)), getKey(map)); // Purpur - getKey
+     }
+ 
+-    // Paper start
++    // Paper start // Purpur start - add key
+     @NotNull
+     public PotionEffect withType(@NotNull PotionEffectType type) {
+-        return new PotionEffect(type, duration, amplifier, ambient, particles, icon);
++        return new PotionEffect(type, duration, amplifier, ambient, particles, icon, key);
```
Comment (Euphillya) [inline patches/api/0045-Potion-NamespacedKey.patch]: Add Purpur comment

File: patches/api/0045-Potion-NamespacedKey.patch [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+     }
+     @NotNull
+     public PotionEffect withDuration(int duration) {
+-        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon);
++        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon, key);
+     }
+     @NotNull
+     public PotionEffect withAmplifier(int amplifier) {
+-        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon);
++        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon, key);
```
Comment (Euphillya) [inline patches/api/0045-Potion-NamespacedKey.patch]: Add Purpur comment

File: patches/api/0045-Potion-NamespacedKey.patch [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
++    // Paper start // Purpur start - add key
+     @NotNull
+     public PotionEffect withType(@NotNull PotionEffectType type) {
+-        return new PotionEffect(type, duration, amplifier, ambient, particles, icon);
++        return new PotionEffect(type, duration, amplifier, ambient, particles, icon, key);
+     }
+     @NotNull
+     public PotionEffect withDuration(int duration) {
+-        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon);
++        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon, key);
```
Comment (Euphillya) [inline patches/api/0045-Potion-NamespacedKey.patch]: Add Purpur comment

File: patches/api/0045-Potion-NamespacedKey.patch [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+     }
+     @NotNull
+     public PotionEffect withAmplifier(int amplifier) {
+-        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon);
++        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon, key);
+     }
+     @NotNull
+     public PotionEffect withAmbient(boolean ambient) {
+-        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon);
++        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon, key);
```
Comment (Euphillya) [inline patches/api/0045-Potion-NamespacedKey.patch]: Add Purpur comment

File: patches/api/0045-Potion-NamespacedKey.patch [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+     }
+     @NotNull
+     public PotionEffect withParticles(boolean particles) {
+-        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon);
++        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon, key);
```
Comment (Euphillya) [inline patches/api/0045-Potion-NamespacedKey.patch]: Add Purpur comment

File: patches/api/0045-Potion-NamespacedKey.patch [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+     }
+     @NotNull
+     public PotionEffect withIcon(boolean icon) {
+-        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon);
++        return new PotionEffect(this.type, duration, amplifier, ambient, particles, icon, key);
```
Comment (Euphillya) [inline patches/api/0045-Potion-NamespacedKey.patch]: Add Purpur Start comment

File: patches/api/0045-Potion-NamespacedKey.patch [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+@@ -98,35 +133,42 @@ public class PotionEffect implements ConfigurationSerializable {
+      * @param map the map to deserialize from
+      */
+     public PotionEffect(@NotNull Map<String, Object> map) {
+-        this(getEffectType(map), getInt(map, DURATION), getInt(map, AMPLIFIER), getBool(map, AMBIENT, false), getBool(map, PARTICLES, true), getBool(map, ICON, getBool(map, PARTICLES, true)));
++        this(getEffectType(map), getInt(map, DURATION), getInt(map, AMPLIFIER), getBool(map, AMBIENT, false), getBool(map, PARTICLES, true), getBool(map, ICON, getBool(map, PARTICLES, true)), getKey(map)); // Purpur - getKey
+     }
+ 
+-    // Paper start
++    // Paper start // Purpur start - add key
```
Comment (Euphillya) [inline patches/api/0045-Potion-NamespacedKey.patch]: dont need comment here

File: patches/api/0045-Potion-NamespacedKey.patch [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+         this.type = type;
+         this.duration = duration;
+@@ -51,8 +56,38 @@ public class PotionEffect implements ConfigurationSerializable {
+         this.ambient = ambient;
+         this.particles = particles;
+         this.icon = icon;
++        this.key = key; // Purpur
++    }
++
++    // Purpur start - Key instead of icon constructor and keyless normal.
```
Comment (Euphillya) [inline patches/api/0045-Potion-NamespacedKey.patch]: Move Purpur Start

File: patches/api/0045-Potion-NamespacedKey.patch [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+     private static final String ICON = "has-icon";
++    private static final String KEY = "namespacedKey"; // Purpur
+     private final int amplifier;
+     private final int duration;
+     private final PotionEffectType type;
+     private final boolean ambient;
+     private final boolean particles;
+     private final boolean icon;
++    @Nullable
++    private final NamespacedKey key; // Purpur
```
Comment (Euphillya) [inline patches/api/0045-Potion-NamespacedKey.patch]: On one line

File: patches/server/0257-Potion-NamespacedKey.patch [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+ import net.minecraft.world.effect.MobEffectInstance;
+ import net.minecraft.world.item.alchemy.Potion;
++import org.bukkit.NamespacedKey;
+ import org.bukkit.support.AbstractTestingBase;
+ import org.junit.Test;
+ 
+@@ -32,6 +33,27 @@ public class PotionTest extends AbstractTestingBase {
+         assertEquals(effects.entrySet().size(), PotionType.values().length - /* PotionTypes with no/shared Effects */ 6);
+     }
+
```
Comment (Euphillya) [inline patches/server/0257-Potion-NamespacedKey.patch:276]: Add Purpur Start Comment

File: patches/server/0257-Potion-NamespacedKey.patch [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
++        assertTrue(namedSpacedEffect.withKey(key).hasKey());
++
++        Map<String, Object> s1 = namedSpacedEffect.serialize();
++        Map<String, Object> s2 = effect.serialize();
++        assertTrue(s1.containsKey("namespacedKey"));
++        assertFalse(s2.containsKey("namespacedKey"));
++        assertNotNull(new PotionEffect(s1).getKey());
++        assertNull(new PotionEffect(s2).getKey());
++    }
++
```
Comment (Euphillya) [inline patches/server/0257-Potion-NamespacedKey.patch:299]: Add Purpur End Comment

**Review by Euphillya — CHANGES_REQUESTED** [review]

**Comment by DaRacci** [comment]
Should I reapply this to the 1.18 branch and make a pull request from there as it seems the 1.17.1 branch has had its final updates

**Comment by encode42** [comment]
Once our 1.18 release is ready, yes. We're currently in a feature freeze as the update takes place

**Comment by encode42** [comment]
Superseded by #783

**Pull request closed without merging**